pydough/conversion/join_key_substitution.py

```diff
@@ -0,0 +1,101 @@
+"""
+Logic for switching references to join keys from one side of a join to the other
+when certain conditions are met, thus allowing the join to be removed by the
+column pruner.
+"""
+
+from pydough.relational import (
+    ColumnReference,
+    ColumnReferenceFinder,
+    Join,
+    JoinCardinality,
+    JoinType,
+    RelationalExpression,
+    RelationalNode,
+    RelationalShuttle,
+)
+from pydough.relational.rel_util import (
+    apply_substitution,
+    extract_equijoin_keys,
+)
+
+
+class JoinKeySubstitutionShuttle(RelationalShuttle):
+    """
+    The relational shuttle that performs join key substitution optimization.
+    """
+
+    def visit_join(self, join: Join) -> RelationalNode:
+        # Build up a mapping of join key substitutions mapping input columns
+        # from one input to another when the optimization case is detected:
+        # requires an inner join with equi-join keys.
+        join_substitution: dict[RelationalExpression, RelationalExpression] = {}
+        if join.join_type == JoinType.INNER:
+            lhs_keys_list, rhs_keys_list = extract_equijoin_keys(join)
+            if len(lhs_keys_list) > 0 and len(rhs_keys_list) > 0:
+                # Identify which columns are used by the join columns that come
+                # from the left and right inputs.
+                lhs_keys: set[ColumnReference] = set(lhs_keys_list)
+                rhs_keys: set[ColumnReference] = set(rhs_keys_list)
+                col_finder = ColumnReferenceFinder()
+                for value in join.columns.values():
+                    value.accept(col_finder)
+                col_refs: set[ColumnReference] = col_finder.get_column_references()
+                lhs_refs = {
+                    ref
+                    for ref in col_refs
+                    if ref.input_name == join.default_input_aliases[0]
```

Comment:
For my understanding, how the `default_input_aliases` identify LHS refs?
I thought it includes both LHS and RHS inputs